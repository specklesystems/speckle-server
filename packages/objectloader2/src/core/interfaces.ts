import Queue from '../queues/queue.js'
import { Item } from '../types/types.js'

export interface Downloader extends Queue<string> {
  initialize(params: {
    results: Queue<Item>
    total: number
    maxDownloadBatchWait?: number
  }): void
  downloadSingle(): Promise<Item>
  disposeAsync(): Promise<void>
}

export interface Database {
  getAll(keys: string[]): Promise<(Item | undefined)[]>
  saveBatch(params: { batch: Item[] }): Promise<void>
  disposeAsync(): Promise<void>
}
