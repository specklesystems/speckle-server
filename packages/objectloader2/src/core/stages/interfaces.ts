import Queue from '../../queues/queue.js'
import { Base, Item } from '../../types/types.js'

export interface Reader {
  initializeQueue(foundQueue: Queue<Item>, notFoundQueue: Queue<string>): void
  getObject(params: { id: string }): Promise<Base>
  requestItem(id: string): void
  requestAll(keys: string[]): void
  disposeAsync(): Promise<void>
}
