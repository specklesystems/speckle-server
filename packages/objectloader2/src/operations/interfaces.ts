import Queue from '../helpers/queue.js'
import { Item } from '../types/types.js'

export interface Downloader extends Queue<string> {
  initializePool(params: {
    results: Queue<Item>
    total: number
    maxDownloadBatchWait?: number
  }): void
  downloadSingle(): Promise<Item>
  disposeAsync(): Promise<void>
}

export interface Database {
  getAll(keys: string[]): Promise<(Item | undefined)[]>
  getItem(params: { id: string }): Promise<Item | undefined>
  cacheSaveBatch(params: { batch: Item[] }): Promise<void>
  disposeAsync(): Promise<void>
}
