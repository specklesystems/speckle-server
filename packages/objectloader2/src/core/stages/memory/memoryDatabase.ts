import { Base, Item } from '../../../types/types.js'
import { Database } from '../../interfaces.js'
import { MemoryDatabaseOptions } from '../../options.js'

export class MemoryDatabase implements Database {
  private items: Map<string, Base>

  constructor(options?: MemoryDatabaseOptions) {
    this.items = options?.items || new Map<string, Base>()
  }

  getAll(keys: string[]): Promise<(Item | undefined)[]> {
    const found: (Item | undefined)[] = []
    for (const key of keys) {
      const item = this.items.get(key)
      if (item) {
        found.push({ baseId: key, base: item })
      } else {
        found.push(undefined)
      }
    }
    return Promise.resolve(found)
  }

  saveBatch({ batch }: { batch: Item[] }): Promise<void> {
    for (const item of batch) {
      if (!item.baseId || !item.base) {
        throw new Error('Item must have a baseId and base')
      }
      this.items.set(item.baseId, item.base)
    }
    return Promise.resolve()
  }

  disposeAsync(): Promise<void> {
    return Promise.resolve()
  }
}
