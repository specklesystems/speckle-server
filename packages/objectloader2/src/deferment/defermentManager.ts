import { DeferredBase } from './deferredBase.js'
import { CustomLogger } from '../types/functions.js'
import { Item, Base } from '../types/types.js'
import { BaseCache } from './BaseCache.js'

export class DefermentManager {
  private deferments: Map<string, DeferredBase> = new Map()
  private logger: CustomLogger
  private disposed = false
  private cache: BaseCache

  constructor(cache: BaseCache, logger: CustomLogger) {
    this.cache = cache
    this.logger = logger
    }

    private now(): number {
      return Date.now()
    }

  get(id: string): DeferredBase | undefined {
    if (this.disposed) throw new Error('DefermentManager is disposed')
    return this.deferments.get(id)
  }

  defer(params: { id: string }): [Promise<Base>, boolean] {
    if (this.disposed) throw new Error('DefermentManager is disposed')
    const item = this.cache.get(params.id)
    if (item) {
      return [Promise.resolve(item.base!), true]
    }
    const deferredBase = this.deferments.get(params.id)
    if (deferredBase) {
      return [deferredBase.getPromise(), true]
    }
    const notYetFound = new DeferredBase(
      params.id
    )
    this.deferments.set(params.id, notYetFound)
    return [notYetFound.getPromise(), false]
  }

  undefer(item: Item, requestItem: (id: string) => void): void {
    if (this.disposed) throw new Error('DefermentManager is disposed')
    const base = item.base
    if (!base) {
      this.logger('undefer called with no base', item)
      return
    }
    this.cache.add(item, (id) => {
      if (!this.deferments.has(id)) {
        requestItem(id)
      }
    })

    //order matters here with found before undefer
    const deferredBase = this.deferments.get(item.baseId)
    if (deferredBase) {
      deferredBase.found(base, item.size || 0)
      this.deferments.delete(item.baseId)
    }
  }

  dispose(): void {
    if (this.disposed) return
    this.disposed = true
    this.clearDeferments()
  }

  private clearDeferments(): void {
    let waiting = 0
    for (const deferredBase of this.deferments.values()) {

      if (deferredBase.getBase() === undefined) {
        waiting++
      }
    }
    this.deferments.clear()
    this.logger('cleared deferments, left', waiting)
  }
}
