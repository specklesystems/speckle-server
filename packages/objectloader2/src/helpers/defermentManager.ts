import { DeferredBase } from './deferredBase.js'
import { Base, CustomLogger, Item } from '../types/types.js'
import { DefermentManagerOptions } from '../operations/options.js'

export class DefermentManager {
  private deferments: Map<string, DeferredBase> = new Map()
  private timer?: ReturnType<typeof setTimeout>
  private logger: CustomLogger
  private currentSize = 0
  private disposed = false

  constructor(private options: DefermentManagerOptions) {
    this.resetGlobalTimer()
    this.logger = options.logger || ((): void => {})
  }

  private now(): number {
    return Date.now()
  }

  isDeferred(id: string): boolean {
    return this.deferments.has(id)
  }

  get(id: string): DeferredBase | undefined {
    if (this.disposed) throw new Error('DefermentManager is disposed')
    return this.deferments.get(id)
  }

  async defer(params: { id: string }): Promise<Base> {
    if (this.disposed) throw new Error('DefermentManager is disposed')
    const now = this.now()
    const deferredBase = this.deferments.get(params.id)
    if (deferredBase) {
      deferredBase.setAccess(now)
      return deferredBase.getPromise()
    }
    const notYetFound = new DeferredBase(
      this.options.ttlms,
      params.id,
      now + this.options.ttlms
    )
    this.deferments.set(params.id, notYetFound)
    return notYetFound.getPromise()
  }

  undefer(item: Item): void {
    if (this.disposed) throw new Error('DefermentManager is disposed')
    const now = this.now()
    this.currentSize += item.size || 0
    //order matters here with found before undefer
    const deferredBase = this.deferments.get(item.baseId)
    if (deferredBase) {
      deferredBase.found(item)
      deferredBase.setAccess(now)
    } else {
      const existing = new DeferredBase(this.options.ttlms, item.baseId, now)
      existing.found(item)
      this.deferments.set(item.baseId, existing)
    }
  }

  private resetGlobalTimer(): void {
    const run = (): void => {
      this.cleanDeferments()
      this.timer = setTimeout(run, this.options.ttlms)
    }
    this.timer = setTimeout(run, this.options.ttlms)
  }

  dispose(): void {
    if (this.disposed) return
    this.disposed = true
    if (this.timer) {
      clearTimeout(this.timer)
      this.timer = undefined
    }
    this.clearDeferments()
  }

  private clearDeferments(): void {
    let waiting = 0
    for (const deferredBase of this.deferments.values()) {
      deferredBase.done(0)
      if (deferredBase.getItem() === undefined) {
        waiting++
      }
    }
    this.currentSize = 0
    this.deferments.clear()
    this.logger('cleared deferments, left', waiting)
  }

  private cleanDeferments(): void {
    const maxSizeBytes = this.options.maxSizeInMb * 1024 * 1024
    if (this.currentSize < maxSizeBytes) {
      this.logger(
        'deferments size is ok, no need to clean',
        this.currentSize,
        maxSizeBytes
      )
      return
    }
    const now = this.now()
    let cleaned = 0
    const start = performance.now()
    for (const deferredBase of Array.from(this.deferments.values())
      .filter((x) => x.isExpired(now))
      .sort((a, b) => this.compareMaybeBasesBySize(a.getItem(), b.getItem()))) {
      if (deferredBase.done(now)) {
        this.currentSize -= deferredBase.getItem()?.size || 0
        this.deferments.delete(deferredBase.getId())
        cleaned++
        if (this.currentSize < maxSizeBytes) {
          break
        }
      }
    }
    this.logger(
      'cleaned deferments, cleaned, left',
      cleaned,
      this.deferments.size,
      performance.now() - start
    )
    return
  }

  compareMaybeBasesBySize(a: Item | undefined, b: Item | undefined): number {
    if (a === undefined && b === undefined) return 0
    if (a === undefined) return -1
    if (b === undefined) return 1
    return this.compareMaybe(a.size, b.size)
  }

  compareMaybe(a: number | undefined, b: number | undefined): number {
    if (a === undefined && b === undefined) return 0
    if (a === undefined) return -1
    if (b === undefined) return 1
    return a - b
  }
}
