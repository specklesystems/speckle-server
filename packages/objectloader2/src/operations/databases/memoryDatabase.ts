import { Base, Item } from '../../types/types.js'
import { Database } from '../interfaces.js'
import { MemoryDatabaseOptions } from '../options.js'

export class MemoryDatabase implements Database {
  private items: Map<string, Base>

  constructor(options?: MemoryDatabaseOptions) {
    this.items = options?.items || new Map<string, Base>()
  }

  getAll(keys: string[]): Promise<(Item | undefined)[]> {
    const found: (Item | undefined)[] = []
    for (const key of keys) {
      const item = this.items.get(key)
      if (item) {
        found.push({ baseId: key, base: item })
      } else {
        found.push(undefined)
      }
    }
    return Promise.resolve(found)
  }

  cacheSaveBatch({ batch }: { batch: Item[] }): Promise<void> {
    for (const item of batch) {
      this.items.set(item.baseId, item.base)
    }
    return Promise.resolve()
  }

  getItem(params: { id: string }): Promise<Item | undefined> {
    const item = this.items.get(params.id)
    if (item) {
      return Promise.resolve({ baseId: params.id, base: item })
    }
    return Promise.resolve(undefined)
  }

  disposeAsync(): Promise<void> {
    return Promise.resolve()
  }
}
