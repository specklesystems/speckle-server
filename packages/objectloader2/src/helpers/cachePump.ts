import { TIME } from '@speckle/shared'
import { Database } from '../operations/interfaces.js'
import { CacheOptions } from '../operations/options.js'
import { CustomLogger, Item } from '../types/types.js'
import BatchingQueue from './batchingQueue.js'
import Queue from './queue.js'
import { Downloader } from '../operations/interfaces.js'
import { DefermentManager } from './defermentManager.js'
import AsyncGeneratorQueue from './asyncGeneratorQueue.js'
import { Pump } from './pump.js'

export class CachePump implements Pump {
  #writeQueue: BatchingQueue<Item> | undefined
  #database: Database
  #logger: CustomLogger
  #deferments: DefermentManager

  #gathered: AsyncGeneratorQueue<Item>

  #options: CacheOptions

  constructor(
    database: Database,
    gathered: AsyncGeneratorQueue<Item>,
    deferments: DefermentManager,
    options: CacheOptions
  ) {
    this.#database = database
    this.#gathered = gathered
    this.#deferments = deferments
    this.#options = options
    this.#logger = options.logger || ((): void => {})
  }

  add(item: Item): void {
    if (!this.#writeQueue) {
      this.#writeQueue = new BatchingQueue({
        batchSize: this.#options.maxCacheWriteSize,
        maxWaitTime: this.#options.maxCacheBatchWriteWait,
        processFunction: (batch: Item[]): Promise<void> =>
          this.#database.cacheSaveBatch({ batch })
      })
    }
    this.#writeQueue.add(item.baseId, item)
  }

  async disposeAsync(): Promise<void> {
    await this.#writeQueue?.disposeAsync()
  }

  async pumpItems(params: {
    ids: string[]
    foundItems: Queue<Item>
    notFoundItems: Queue<string>
  }): Promise<void> {
    const { ids, foundItems, notFoundItems } = params
    const maxCacheReadSize = this.#options.maxCacheReadSize

    for (let i = 0; i < ids.length; ) {
      if (this.#writeQueue?.isDisposed()) break
      if ((this.#writeQueue?.count() ?? 0) > this.#options.maxWriteQueueSize) {
        this.#logger(
          'pausing reads (# in write queue: ' + this.#writeQueue?.count() + ')'
        )
        await new Promise((resolve) => setTimeout(resolve, TIME.second)) // Pause for 1 second, protects against out of memory
        continue
      }
      const batch = ids.slice(i, i + maxCacheReadSize)
      const cachedData = await this.#database.getAll(batch)
      for (let i = 0; i < cachedData.length; i++) {
        if (cachedData[i]) {
          foundItems.add(cachedData[i]!)
        } else {
          notFoundItems.add(batch[i])
        }
      }
      i += maxCacheReadSize
    }
  }

  async *gather(ids: string[], downloader: Downloader): AsyncGenerator<Item> {
    const total = ids.length
    const pumpPromise = this.pumpItems({
      ids,
      foundItems: this.#gathered,
      notFoundItems: downloader
    })
    let count = 0
    for await (const item of this.#gathered.consume()) {
      this.#deferments.undefer(item)
      yield item
      count++
      if (count >= total) {
        this.#gathered.dispose()
      }
    }
    await pumpPromise
  }
}
