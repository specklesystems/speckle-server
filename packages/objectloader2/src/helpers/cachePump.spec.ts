import { describe, expect, test } from 'vitest'
import { CachePump } from './cachePump.js'
import { Base, Item } from '../types/types.js'
import BufferQueue from './bufferQueue.js'
import AsyncGeneratorQueue from './asyncGeneratorQueue.js'
import { DefermentManager } from './defermentManager.js'
import { MemoryDatabase } from '../operations/databases/memoryDatabase.js'
import { Database } from '../operations/interfaces.js'

describe('CachePump testing', () => {
  test('write two items to queue use pumpItems that are NOT found', async () => {
    const i1: Item = { baseId: 'id1', base: { id: 'id', speckle_type: 'type' } }
    const i2: Item = { baseId: 'id2', base: { id: 'id', speckle_type: 'type' } }

    const gathered = new AsyncGeneratorQueue<Item>()
    const deferments = new DefermentManager({ maxSizeInMb: 1, ttlms: 1 })
    const cachePump = new CachePump(new MemoryDatabase({}), gathered, deferments, {
      maxCacheReadSize: 1,
      maxCacheWriteSize: 1,
      maxCacheBatchWriteWait: 1,
      maxCacheBatchReadWait: 1,
      maxWriteQueueSize: 1
    })

    const foundItems = new BufferQueue<Item>()
    const notFoundItems = new BufferQueue<string>()

    await cachePump.pumpItems({
      ids: [i1.baseId, i2.baseId],
      foundItems,
      notFoundItems
    })

    expect(foundItems.values()).toMatchSnapshot()
    expect(notFoundItems.values()).toMatchSnapshot()
    await cachePump.disposeAsync()
  })

  test('write two items to queue use pumpItems that are found', async () => {
    const i1: Item = { baseId: 'id1', base: { id: 'id', speckle_type: 'type' } }
    const i2: Item = { baseId: 'id2', base: { id: 'id', speckle_type: 'type' } }

    const db = new Map<string, Base>()
    db.set(i1.baseId, i1.base)
    db.set(i2.baseId, i2.base)

    const gathered = new AsyncGeneratorQueue<Item>()
    const deferments = new DefermentManager({ maxSizeInMb: 1, ttlms: 1 })
    const cachePump = new CachePump(
      new MemoryDatabase({ items: db }),
      gathered,
      deferments,
      {
        maxCacheReadSize: 1,
        maxCacheWriteSize: 1,
        maxCacheBatchWriteWait: 1,
        maxCacheBatchReadWait: 1,
        maxWriteQueueSize: 1
      }
    )

    const foundItems = new BufferQueue<Item>()
    const notFoundItems = new BufferQueue<string>()

    await cachePump.pumpItems({
      ids: [i1.baseId, i2.baseId],
      foundItems,
      notFoundItems
    })

    expect(foundItems.values()).toMatchSnapshot()
    expect(notFoundItems.values()).toMatchSnapshot()
    await cachePump.disposeAsync()
  })

  test('can dispose while waiting and not wait', async () => {
    const i1: Item = { baseId: 'id1', base: { id: 'id', speckle_type: 'type' } }
    const i2: Item = { baseId: 'id2', base: { id: 'id', speckle_type: 'type' } }

    const db: Database = {
      getAll: async () => Promise.resolve([])
    } as unknown as Database
    const gathered = new AsyncGeneratorQueue<Item>()
    const deferments = new DefermentManager({ maxSizeInMb: 1, ttlms: 1 })
    const cachePump = new CachePump(db, gathered, deferments, {
      maxCacheReadSize: 1,
      maxCacheWriteSize: 1,
      maxCacheBatchWriteWait: 1,
      maxCacheBatchReadWait: 1,
      maxWriteQueueSize: 1
    })

    const foundItems = new BufferQueue<Item>()
    const notFoundItems = new BufferQueue<string>()

    await cachePump.disposeAsync()
    await cachePump.pumpItems({
      ids: [i1.baseId, i2.baseId],
      foundItems,
      notFoundItems
    })
  })
})
