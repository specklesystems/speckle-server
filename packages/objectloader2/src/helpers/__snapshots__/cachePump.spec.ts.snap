// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CachePump testing > write two items to queue use pumpItems that are NOT found 1`] = `[]`;

exports[`CachePump testing > write two items to queue use pumpItems that are NOT found 2`] = `
[
  "id1",
  "id2",
]
`;

exports[`CachePump testing > write two items to queue use pumpItems that are found 1`] = `
[
  {
    "base": {
      "id": "id",
      "speckle_type": "type",
    },
    "baseId": "id1",
  },
  {
    "base": {
      "id": "id",
      "speckle_type": "type",
    },
    "baseId": "id2",
  },
]
`;

exports[`CachePump testing > write two items to queue use pumpItems that are found 2`] = `[]`;
